// mangadex-api/dist/lib/normalize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeChapter = exports.normalizeManga = void 0;
const Composer_1 = require("../Composer");
const normalizeManga = (manga) => ({
    ...manga,
    links: Composer_1.Composer.getMangaLinks(manga.links),
    description: Composer_1.Composer.parseToHTML(manga.description),
    publication: {
        ...manga.publication,
        languageName: Composer_1.Composer.getLangName(manga.publication.language)
    }
});
exports.normalizeManga = normalizeManga;
const normalizeChapter = (chapter) => ({
    ...chapter,
    pages: Array.isArray(chapter.pages)
        ? chapter.pages.map((page) => `${chapter.server}${chapter.hash}/${page}`)
        : chapter.pages,
    fallbackPages: Array.isArray(chapter.pages)
        ? chapter.pages.map((page) => `${chapter.serverFallback || chapter.server}${chapter.hash}/${page}`)
        : null,
    languageName: Composer_1.Composer.getLangName(chapter.language)
});
exports.normalizeChapter = normalizeChapter;
