// mangadex-api/dist/Composer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Composer = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = require("path");
const links_1 = require("./lib/links");
const langCodeData = JSON.parse(fs_1.default.readFileSync(path_1.join(__dirname, '..', 'langcodes.json'), 'utf-8'));
class Composer {
    static getMangaLinks(links) {
        return Object.entries(links).map(([name, value]) => {
            const matchedLink = links_1.links[name];
            return {
                title: matchedLink ? matchedLink.title : name,
                url: matchedLink ? matchedLink.url + value : value
            };
        });
    }
    static getLangName(langCode) {
        return langCodeData[langCode] || langCode;
    }
    static parseToHTML(string) {
        return string.replace(/\[url=(\S+?)\](\S+?)\[\/url\]/gi, '<a href="$1">$2</a>');
    }
    static formatTypeMapToArray(map) {
        return Object.values(map);
    }
}
exports.Composer = Composer;
